### DISCUSSION TITLE
using fetch()

### DISCUSSION THREAD

**Discussion by BobSquarePants** [brython-dev/brython#2176] [Q&A] [answered]
Hi everyone,

I would like to use an online API than return a JSON.. 
<details><summary>out of topic</summary>
I rather preferred had a YAML reply.. but it's not provided yet..
</details> 

The requests to the API work with a GET to an URL with `(?key=value)`
example `https://theAPI.com?key=value`

I would like to process the data in the frontend. 
I've only found the[ fetch() method..](https://developer.mozilla.org/en-US/docs/Web/API/fetch)  (maybe something more simpler exist ??? )

The code readability on the example is indecipherable, not the fault of MDN but because of the JS nature syntax , And no matter what, I can't put my head around it..

So how could we use `fetch()` trough Bryhton (without _stdlib.js )

Thanks.

**Comment by BobSquarePants** [comment]
I've tried this

```python
def f_succeed():
    print('succeed')
	
window.fetch(window.encodeURI('https://theAPI.com?test=test'), {'redirect':'error'}).then(f_succeed)
```
but of course it's not working... 😢 

error
> Uncaught (in promise) Error:

**Comment by PierreQuentel** [comment]
Requests sent from a browser can only reach resources with the [same origin](https://developer.mozilla.org/en-US/docs/Web/Security/Same-origin_policy) as the calling script, except if the server where the resource stands sets the cross-origin headers. Is it the case in your example ?

If so, you can use the Javascript-like syntax `window.fetch(url).then(handle_reponse)`, or a maybe more pythonic async code like

```python
from browser import window, aio

async def f():
    response = await window.fetch('test.html')
    if response.status == 200:
        text = await response.text()
        print(text)

aio.run(f())
```

**Reply by BobSquarePants** [reply]
Hi @PierreQuentel ,

Once again thank you for your guidance.

I will start first with a JS-like syntax 👎, then later on with the extra module aio.

I have now

```python
def fetched(ResJSobj):
    
    if ResJSobj.ok == True:
        print(ResJSobj.text()) 
    else:
        print("it didn't worked")

window.fetch('/aLocalPage').then(fetched)
```
but the successful `print` return
> <Javascript object: [object Promise]>

is there a way to get the text string (without aio module ? )

**Comment by PierreQuentel** [comment, marked as answer]
As the [documentation of fetch()](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) says, it resolves to a `Response` object, and `Response.text()` is a promise that resolves to the text content of the resource. You should do something like
```python
def handle_content(s):
    print(s)
    
def fetched(ResJSobj):
    if ResJSobj.ok:
        ResJSobj.text().then(handle_content)
    else:
        print("it didn't work")

window.fetch('/aLocalPage').then(fetched)
```

**Reply by BobSquarePants** [reply]
Thank you @PierreQuentel  🙏🏽 !

I didn't fully understood that `then()` was a callback caller.

It's not Pytonic.. but it will do until I manage to get AIO up and running :)
